Implement Disassembly View and InstructionBreakpoints

colin-grant-work opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

> Leaving this as a draft for now, as there are some details I need to check to make sure instruction breakpoints are working correctly.

Addresses parts of #8136 by implementing InstructionBreakpoints and a disassembly view to and the return of the disassemble request.

This PR copies and adapts VSCode's implementation of the Disassembly View for Theia.

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

0. In general: use any debugger that supports the disassemble request, and confirm that the behavior in Theia is consistent with the behavior in VSCode.
1. For example...
2. Install the [CDT-GDB VSCode plugin](https://open-vsx.org/extension/eclipse-cdt/cdt-gdb-vscode)
3. Open [this example workspace](https://github.com/eclipse-theia/theia-cpp-extensions/tree/master/examples/cpp-debug-workspace) and execute the build task.
4. Set a breakpoint in `a.cpp` and start debugging.
5. When you hit a breakpoint, you should be able to open the context menu in the stack view and find the 'Open Disassembly View' command - it should also be available in the command palette.
6. Disassembly should load.
7. Scrolling should load more lines.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

msujew reviewed (commented):
Looks mostly good. I noticed some unexpected behavior in regards to disassembly breakpoints:

![image](https://user-images.githubusercontent.com/4377073/174640546-622712ed-367e-46a0-b89f-a48b17d54a5e.png)

As far as I can tell, debugging directly in the disassembly view is not supported, right (that's what the icon indicates)? Anyway, these breakpoints cannot be removed, and clicking on the breakpoint sidebar in the disassembly view only adds more breakpoints, instead of removing the existing ones.

vince-fugnitto commented:
The CQ has been approved, thank you @JonasHelming for helping it get across the line :+1:

colin-grant-work commented:
@msujew, I pushed a change (and meant to mention it - sorry) that addresses your comments. I have modified the enablement of the commands that manipulate all breakpoints to be available when only instruction breakpoints are present. I have also (and this makes it harder to test my last claim ;-)) added code that prevents the creation of instruction breakpoints for sessions that don't support them - an improvement, I believe, over VSCode's code.

msujew reviewed (approved):
Great, looks good to me!

- [x] Hitting a breakpoint in the C code lets us open the disassembly view via a context menu / command.
- [x] Scrolling up/down in the disassembly view loads more instructions and (relatively accurate) changes the position within the list view.
- [x] Breakpoints within the disassembly view are disabled for `cdt-gdb`, definitely an improvements compared to vscode.

We can't really test the functionality of breakpoints within the disassembly view, but I'm fine with revisiting this if there are any problems with it and some extension has mature support for it.

colin-grant-work commented on packages/debug/src/browser/disassembly-view/disassembly-view-widget.ts:
```diff
@@ -0,0 +1,465 @@
+            return false;
+        }
+
+        // console.log(`DisassemblyView: loadDisassembledInstructions ${memoryReference}, ${instructionOffset}, ${instructionCount}`);
```
The answer is, 'yes, probably,' but it's actually in the VSCode source, so I didn't remove it; no very good reason to keep it, though.
